Statements: Parsing & dependency management

neenaoffline opened the pull request:
This PR adds statement parsing & some dependency management. Statement evaluation will be added in a follow-up PR.

prabhanshuguptagit commented on src/bean/grid.cljs:
```diff
@@ -156,32 +145,41 @@
+               sheet
+               (->> updated-addrs
+                    (deps/immediate-dependents sheet)
+                    (deps/resolve-dependents sheet)))))))
```
Would passing `updated-addrs` to `resolve-dependents` directly take care of getting the immediate dependents?

prabhanshuguptagit commented on src/bean/interpreter.cljs:
```diff
@@ -153,13 +117,31 @@
+  (-> (eval-ast (:ast cell) sheet)
       (ast-result->cell cell)))
+
+(defn eval-statements [scratch]
+  (comment
+    eval scratch
+    generate bindings names -> ast))
```
Can remove.

prabhanshuguptagit commented on src/bean/deps.cljs:
```diff
@@ -0,0 +1,84 @@
+      #{})))
+
+(defn resolve-dependents [{:keys [depgraph]} deps]
+  "Iterate over the dependents and return a set of resolved dependents. :ref
```
Could break this into two `resolve-binding-dependents`, then this could just be called `dependents`.

neenaoffline committed:
Remove run-all-tests from provenance test ns

neenaoffline committed:
Explicit sheet initialization

neenaoffline committed:
Extract deps code into deps ns

neenaoffline committed:
Make address the first arg of eval-sheet when addr exists

neenaoffline committed:
Separate eval-address from eval-sheet

neenaoffline committed:
Pass address type to eval-address

neenaoffline committed:
eval-address for :named addresses

neenaoffline committed:
Add separate deps tests

neenaoffline committed:
Tie in secondary space UI with the functionality

neenaoffline committed:
Handle errors in statement handling

prabhanshuguptagit committed:
Add buttons to switch between dark/light mode

prabhanshuguptagit commented on src/bean/deps.cljs:
```diff
@@ -1,4 +1,56 @@
 (defn ->ref-dep [dep]
-  [:ref dep])
\ No newline at end of file
+  [:cell dep])
```
`:CellRef`?

prabhanshuguptagit commented on src/bean/ui/scratch.cljs:
```diff
@@ -0,0 +1,46 @@
+    [:button {:class [:scratch-header-btn]
+              :on-click (fn [_]
+                          (swap! sheet #(-> %
+                                            scratch/reevaluate
```
Should also re-evaluate dependent cells when code is evaluated. Might be better to do it via the scratch re-evaluation?

neenaoffline committed:
Add content to statements. Change dep type from :Name -> :named.

To add content to statemnents, we annotate the AST with the source using
instaparse's built-in functionality.

prabhanshuguptagit commented on src/bean/ui/scratch.cljs:
```diff
@@ -0,0 +1,46 @@
+(ns bean.ui.scratch
+  (:require [bean.scratch :as scratch]
```
Rename scratch ns to `code`.

prabhanshuguptagit commented on src/bean/deps.cljs:
```diff
@@ -1,4 +1,56 @@
+  (:require [clojure.set :as set]
+            [bean.util :as util]))
 
 (defn ->ref-dep [dep]
```
```suggestion
(defn ->cell-dep [dep]
```

prabhanshuguptagit commented on src/bean/deps.cljs:
```diff
@@ -1,4 +1,56 @@
+      :FunctionInvocation (apply set/union
+                                 (map ast->deps args))
+      :FunctionDefinition (ast->deps arg)
+      :Name (if (#{"x" "y" "z"} arg) #{} #{(->named-dep arg)})
```
Move this check to a `formal-parameter?` in interpreter.

prabhanshuguptagit commented on src/bean/deps.cljs:
```diff
@@ -1,4 +1,56 @@
+(defn- depgraph-remove-edge [depgraph parent child]
+  (let [updated-dependent-set (disj (get depgraph parent) child)]
+    (if (empty? updated-dependent-set)
+      (dissoc depgraph parent)
```
Why do we need to dissoc parents with no children?

prabhanshuguptagit commented on src/bean/grid.cljs:
```diff
@@ -144,63 +120,135 @@
 
-  ([{:keys [grid] :as sheet} address]
-   (eval-sheet sheet address (util/get-cell grid address) false))
+(defmulti eval-address first)
```
question for self: how does this get the the `first` of the first argument.

prabhanshuguptagit commented on src/bean/parser.cljs:
```diff
@@ -35,8 +35,17 @@
+(defn error [result]
+  (when (insta/get-failure result)
+    (let [{:keys [index reason]} result]
+      (str "Parse Error: idx " index ". " reason))))
```
note to self: how is this used.

prabhanshuguptagit commented on src/bean/parser.cljs:
```diff
@@ -35,8 +35,17 @@
-  (insta/parse statement-parser v))
+(defn parse-statement [src]
+  (let [program (insta/parse statement-parser src)]
+    (insta/add-line-and-column-info-to-metadata src program)))
```
Should add a test for this.

prabhanshuguptagit commented on src/bean/scratch.cljs:
```diff
@@ -0,0 +1,11 @@
+(ns bean.scratch
```
Rename ns.

prabhanshuguptagit commented on src/bean/grid.cljs:
```diff
@@ -144,63 +120,135 @@
+     (errors/undefined-named-ref named)))
+
+  ([address sheet new-content]
+  ;; TODO: This is invalid
```
?

prabhanshuguptagit commented on src/bean/grid.cljs:
```diff
@@ -134,7 +110,7 @@
 
 (defn parse-grid [grid]
-  (util/map-on-matrix content->cell grid))
+  (util/map-on-matrix value/from-cell grid))
```
This name seems incorrect, it's more like `to-cell`. This is the input to the interpreter, the interpreter returns a `value`. See `ast-result`.

prabhanshuguptagit commented on src/bean/errors.cljs:
```diff
@@ -0,0 +1,8 @@
+   :representation (str "Undefined reference: \"" name "\"")})
+
+(defn named-ref-error [named error]
+  (str "name: " named ". " error))
```
note to self: Look at `named-ref-error` usage.

prabhanshuguptagit commented on test/bean/deps_test.cljs:
```diff
@@ -0,0 +1,41 @@
+            [bean.deps :refer [make-depgraph update-depgraph]]
+            [clojure.test :refer [deftest testing is]]))
+
+(deftest depgraph-update-test
```
Add a test for named deps.

prabhanshuguptagit commented on test/bean/provenance_test.cljs:
```diff
@@ -50,8 +52,8 @@
-    (let [evaled-grid (grid/eval-sheet [["1" "2" "=A1+B1"]
-                                        ["=C1" "" ""]])]
+    (let [evaled-grid (grid/eval-sheet
+                       (grid/new-sheet [["1" "2" "=A1+B1"] ["=C1" "" ""]] ""))]
```
Split grid into two lines if the linter allows. It's easier to read.

prabhanshuguptagit commented on src/bean/ui/scratch.cljs:
```diff
@@ -0,0 +1,46 @@
+
+(defn text-area [sheet]
+  [:div {:class (cs (str "scratch-state-"
+                         (name (or (get-in @sheet [:ui :scratch-evaluation-state])
+                                   :evaluated)))
```
Should put `scratch-evaluation-state ` `evaluated` by default instead of defaulting here.

prabhanshuguptagit commented on src/bean/grid.cljs:
```diff
@@ -144,63 +120,135 @@
-  {:grid parsed-grid
-   :bindings {}
-   :depgraph (make-depgraph parsed-grid)})
+(defn new-sheet [content-grid code]
```
Can make code default to `""` unless supplied.

prabhanshuguptagit commented on src/bean/grid.cljs:
```diff
@@ -144,63 +120,135 @@
-      eval-sheet
-      {:grid grid*
-       :depgraph (cond-> depgraph
-                   content-changed? (update-depgraph
```
Maybe can skip `content-changed?` for simplicity.
